spec/ics-010-grandpa-client/README.md

```diff
@@ -0,0 +1,345 @@
+
+```typescript
+interface Evidence {
+  fromHeight: uint64
+  h1: Header
+  h2: Header
+}
+```
+
+### Client initialisation
+
+GRANDPA client initialisation requires a (subjectively chosen) latest consensus state, including the full authority set.
+
+```typescript
+function initialise(identifier: Identifier, height: uint64, consensusState: ConsensusState): ClientState {
+    set("clients/{identifier}/consensusStates/{height}", consensusState)
+    return ClientState{
+      latestHeight: height,
+      frozenHeight: null,
+    }
+}
+```
+
+The GRANDPA client `latestClientHeight` function returns the latest stored height, which is updated every time a new (more recent) header is validated.
+
+```typescript
+function latestClientHeight(clientState: ClientState): uint64 {
+  return clientState.latestHeight
+}
+```
+
+### Validity predicate
+
+GRANDPA client validity checking verifies a header is signed by the current authority set and verifies the authority set proof to determine if there is a expected change to the authority set. If the provided header is valid, the client state is updated & the newly verified commitment written to the store.
+
+```typescript
+function checkValidityAndUpdateState(
+  clientState: ClientState,
+  header: Header) {
+    // assert header height is newer than any we know
+    assert(header.height > clientState.latestHeight)
+    consensusState = get("clients/{identifier}/consensusStates/{clientState.latestHeight}")
+    // verify that the provided header is valid
+    assert(verify(header.justification, consensusState.authoritySet))
+    // update latest height
+    clientState.latestHeight = header.height
+    // verify that the authority set has been stored
+    assert(header.commitmentRoot.verifyMembership(path, authoritySet, header.authoritySetProof))
+    // create recorded consensus state, save it
+    consensusState = ConsensusState{authoritySet, header.commitmentRoot}
+    set("clients/{identifier}/consensusStates/{header.height}", consensusState)
+    // save the client
+    set("clients/{identifier}", clientState)
+}
+```
+
+### Misbehaviour predicate
+
+GRANDPA client misbehaviour checking determines whether or not two conflicting headers at the same height would have convinced the light client.
+
+```typescript
+function checkMisbehaviourAndUpdateState(
+  clientState: ClientState,
+  evidence: Evidence) {
+    // assert that the heights are the same
+    assert(evidence.h1.height === evidence.h2.height)
+    // assert that the commitments are different
+    assert(evidence.h1.commitmentRoot !== evidence.h2.commitmentRoot)
+    // fetch the previously verified commitment root & validator set
+    consensusState = get("clients/{identifier}/consensusStates/{evidence.fromHeight}")
+    // check if the light client "would have been fooled"
+    assert(
+      verify(consensusState.authoritySet, evidence.fromHeight, evidence.h1) &&
```

Comment:
This `verify` function has different arguments than the one defined above

Reply:
fixed.